### DISCUSSION TITLE
Integration with other dweb protocols

### DISCUSSION THREAD

**Discussion by ple1n** [freenet/freenet-core#163] [General]
Contracts are fully retrieved by all their peers, so they are not meant for heavy data storage, or computation. For example, videos, git repositories, video transcoding. Integration with other protocols should be necessary in the future.

- Bittorrent, which is better in anonymity
- IPFS, more flexible, but immature
- IPFS + Filecoin retrieval market, WIP

How will these protocols be integrated with Locutus ? Compile them into Locutus contracts ? probably not. Some plugin system ? And they are shipped with Locutus ?

Locutus contracts would store static files in IPFS, and the user pin them locally. When super nodes get to know the files they pin them on the servers for caching and upload them to Filecoin for long-term storage. Other users retrieve the files via the cache on IPFS or use a few tokens to pay Filecoin retrieval market for the data. (still unclear how their CDN will work. payment channels are expensive) Considering a PeerTube, a decentralized CDN could be cost-efficient. IPFS is certainly necessary for large files, but Filecoin might be too expensive to use.

Contracts can store large files, but for videos, IPFS can offer the IPLD and its chunking algorithms. And IPFS network has more nodes, which means more cache and locutus can share content with other dweb.

**Comment by sanity** [comment]
Locutus will be able to chunk large files across multiple contracts, which could support efficient seeking in audio/video. The original Freenet did something similar.
